### ISSUE TITLE
DOS attack using string>number

### ISSUE THREAD

**Issue by mrjbq7** [slavapestov/factor#1231]
This is kinda lame performance:

```factor
IN: scratchpad gc [ "1.0e1234567" string>number ] time .
Running time: 9.597006757000001 seconds

IN: scratchpad gc [ 10.0 1234567 ^ ] time
Running time: 1.1161e-05 seconds
```

Obviously the second one just calls ``fpow``, but since ``^`` is implemented in ``math.functions``, we can't use it from ``math.parser``... perhaps there's a way to bootstrap a slow implementation and then when ``math.libm`` is loaded use the fast version (and when ``math.complex`` is loaded, add ``complex^`` support).

**Comment by mrjbq7** [comment]
Made the DOS a little less bad in ccddbebf94d14c12b44999c0ba47ef2d15cf2069 but it's still terrible.

**Comment by jonenst** [comment]
I've got a work in progress solution for this along with other changes to math.parser at  https://github.com/jonenst/factor/tree/dos_numberstring, will create a PR soon if it looks good

**Comment by mrjbq7** [comment]
@jonenst Awesome! Can't wait!

**Comment by jonenst** [comment]
https://github.com/jonenst/factor/tree/dos_numberstring is ready, but I ran into an unrelated bug that's blocking it, will need to investigate:
```
IN: scratchpad 5 [1,b] [ 2 / 2 1024 ^ / >float . ] each
2.781342323134002e-309
-1/0.
NAN: 8000000000000
0.0
5.562684646268003e-309
```

The effect on number parsing after this pr is that those bad values are returned for numbers like 0x1.0-p1024, whereas before the correct value was returned. This is just because of https://github.com/jonenst/factor/commit/445b58fb7cf0ef1e38f9a3369a71bd34d1cd5908, it has nothing to do with huge exponents, but it's a real bug we have to fix I think

**Comment by jonenst** [comment]
Ok, I found the problem in https://github.com/slavapestov/factor/pull/1406; After merge it, this is good to go I think

**Comment by mrjbq7** [comment]
Hi @jonenst I'm a little confused about which branches you want to merge, is there a PR for the fix to this issue or is it just your ``dos_numberstring`` branch?

**Comment by jonenst** [comment]
Ah sorry, I said I would create a PR but forgot about it :) I'll create one.

**Comment by jonenst** [comment]
rebased after your changes on using mutable objects in https://github.com/slavapestov/factor/pull/1408

**Comment by mrjbq7** [comment]
Fixed by #1408!  Thanks @jonenst!
